File: java/src/main/java/frc/engineer/packets/PacketModels.java
package frc.engineer.packets;

import java.util.ArrayList;
import java.util.List;

public final class PacketModels {
    private PacketModels() {
    }

    public static final class PacketHeader {
        public Integer packetFormat;
        public Integer gameYear;
        public Integer gameMajorVersion;
        public Integer gameMinorVersion;
        public Integer packetVersion;
        public Integer packetId;
        public Long sessionUid;
        public Float sessionTime;
        public Long frameIdentifier;
        public Long overallFrameIdentifier;
        public Integer playerCarIndex;
        public Integer secondaryPlayerCarIndex;
    }

    public static final class PacketSessionData {
        public Integer weather;
        public Integer trackTemperature;
        public Integer airTemperature;
        public Integer totalLaps;
        public Integer trackLength;
        public Integer sessionType;
        public Integer trackId;
        public Integer pitSpeedLimit;
    }

    public static final class PacketEventData {
        public String eventStringCode;
        public String details;
    }

    public static final class ParticipantData {
        public Boolean aiControlled;
        public Integer driverId;
        public Integer networkId;
        public Integer teamId;
        public String name;
        public Integer yourTelemetry;
    }

    public static final class PacketParticipantsData {
        public List<ParticipantData> participants = new ArrayList<>();
    }

    public static final class CarMotionData {
        public Float worldPositionX;
        public Float worldPositionY;
        public Float worldPositionZ;
        public Float worldVelocityX;
        public Float worldVelocityY;
        public Float worldVelocityZ;
        public Float gForceLateral;
        public Float gForceLongitudinal;
        public Float gForceVertical;
        public Float yaw;
        public Float pitch;
        public Float roll;
    }

    public static final class PacketMotionData {
        public List<CarMotionData> carMotionData = new ArrayList<>();
    }

    public static final class LapData {
        public Integer position;
        public Integer currentLapNum;
        public Float lapDistance;
        public Float lastLapTime;
        public Float bestLapTime;
        public Integer sector1TimeMs;
        public Integer sector2TimeMs;
        public Integer currentSector;
        public Integer deltaToCarInFront;
        public Integer deltaToRaceLeader;
        public Integer pitStatus;
        public Integer driverStatus;
        public Integer resultStatus;
        public Integer numPitStops;
        public Float pitStopTimer;
        public Integer pitLaneTime;
        public Integer penalties;
        public Integer totalWarnings;
        public Integer cornerCuttingWarnings;
        public Integer numUnservedDriveThroughs;
        public Integer numUnservedStopGoPenalties;
        public Boolean lapInvalid;
    }

    public static final class PacketLapData {
        public List<LapData> lapData = new ArrayList<>();
    }

    public static final class CarSetupData {
        public Integer frontWing;
        public Integer rearWing;
        public Integer onThrottle;
        public Integer offThrottle;
        public Float frontCamber;
        public Float rearCamber;
        public Float frontToe;
        public Float rearToe;
        public Integer frontSuspension;
        public Integer rearSuspension;
        public Integer frontAntiRollBar;
        public Integer rearAntiRollBar;
        public Integer frontSuspensionHeight;
        public Integer rearSuspensionHeight;
        public Integer brakePressure;
        public Integer brakeBias;
        public Float rearLeftTyrePressure;
        public Float rearRightTyrePressure;
        public Float frontLeftTyrePressure;
        public Float frontRightTyrePressure;
        public Integer ballast;
        public Float fuelLoad;
    }

    public static final class PacketCarSetupData {
        public List<CarSetupData> carSetups = new ArrayList<>();
    }

    public static final class CarTelemetryData {
        public Integer speed;
        public Float throttle;
        public Float brake;
        public Float steering;
        public Integer gear;
        public int[] brakeTemp = new int[] {0, 0, 0, 0};
        public int[] tireSurfaceTemp = new int[] {0, 0, 0, 0};
        public int[] tireInnerTemp = new int[] {0, 0, 0, 0};
        public float[] tirePressure = new float[] {0f, 0f, 0f, 0f};
        public int[] surfaceType = new int[] {0, 0, 0, 0};
        public Integer engineRpm;
        public Integer drs;
    }

    public static final class PacketCarTelemetryData {
        public List<CarTelemetryData> carTelemetryData = new ArrayList<>();
    }

    public static final class CarStatusData {
        public Integer tractionControl;
        public Integer antiLockBrakes;
        public Integer fuelMix;
        public Integer brakeBias;
        public Integer pitLimiterStatus;
        public Float fuelInTank;
        public Float fuelCapacity;
        public Float fuelRemainingLaps;
        public Integer maxRpm;
        public Integer idleRpm;
        public Integer maxGears;
        public Integer drsAllowed;
        public Integer actualTyreCompound;
        public Integer visualTyreCompound;
        public Integer tireAge;
        public Integer vehicleFiaFlags;
        public Float ersStoreEnergy;
        public Integer ersDeployMode;
        public Float ersHarvestedMGUK;
        public Float ersHarvestedMGUH;
        public Float ersDeployedThisLap;
    }

    public static final class PacketCarStatusData {
        public List<CarStatusData> carStatusData = new ArrayList<>();
    }

    public static final class CarDamageData {
        public int[] tireWear = new int[] {0, 0, 0, 0};
        public int[] tiresDamage = new int[] {0, 0, 0, 0};
        public int[] brakesDamage = new int[] {0, 0, 0, 0};
        public Integer frontLeftWingDamage;
        public Integer frontRightWingDamage;
        public Integer rearWingDamage;
        public Integer floorDamage;
        public Integer diffuserDamage;
        public Integer sidepodDamage;
        public Integer gearboxDamage;
        public Integer engineDamage;
    }

    public static final class PacketCarDamageData {
        public List<CarDamageData> carDamageData = new ArrayList<>();
    }

    public static final class LapHistoryData {
        public Long lapTimeInMs;
        public Integer sector1TimeInMs;
        public Integer sector2TimeInMs;
        public Integer sector3TimeInMs;
        public Integer lapValidBitFlags;
    }

    public static final class PacketSessionHistoryData {
        public Integer carIndex;
        public Integer numLaps;
        public Integer bestLapTimeLapNum;
        public Integer bestSector1LapNum;
        public Integer bestSector2LapNum;
        public Integer bestSector3LapNum;
        public Integer bestSector1Ms;
        public Integer bestSector2Ms;
        public Integer bestSector3Ms;
        public List<LapHistoryData> lapHistoryData = new ArrayList<>();
    }

    public static final class TyreSetData {
        public Integer actualTyreCompound;
        public Integer visualTyreCompound;
        public Integer wear;
        public Boolean available;
        public Integer recommendedSession;
    }

    public static final class PacketTyreSetsData {
        public Integer carIndex;
        public Integer fittedIdx;
        public List<TyreSetData> availableSets = new ArrayList<>();
    }
}

File: kotlin/src/main/kotlin/frc/engineer/packets/PacketModels.kt
package frc.engineer.packets

data class PacketHeader(
    val packetFormat: Int? = null,
    val gameYear: Int? = null,
    val gameMajorVersion: Int? = null,
    val gameMinorVersion: Int? = null,
    val packetVersion: Int? = null,
    val packetId: Int? = null,
    val sessionUid: Long? = null,
    val sessionTime: Float? = null,
    val frameIdentifier: Long? = null,
    val overallFrameIdentifier: Long? = null,
    val playerCarIndex: Int? = null,
    val secondaryPlayerCarIndex: Int? = null,
)

data class PacketSessionData(
    val weather: Int? = null,
    val trackTemperature: Int? = null,
    val airTemperature: Int? = null,
    val totalLaps: Int? = null,
    val trackLength: Int? = null,
    val sessionType: Int? = null,
    val trackId: Int? = null,
    val pitSpeedLimit: Int? = null,
)

data class PacketEventData(
    val eventStringCode: String? = null,
    val details: String? = null,
)

data class ParticipantData(
    val aiControlled: Boolean? = null,
    val driverId: Int? = null,
    val networkId: Int? = null,
    val teamId: Int? = null,
    val name: String? = null,
    val yourTelemetry: Int? = null,
)

data class PacketParticipantsData(
    val participants: List<ParticipantData> = emptyList(),
)

data class CarMotionData(
    val worldPositionX: Float? = null,
    val worldPositionY: Float? = null,
    val worldPositionZ: Float? = null,
    val worldVelocityX: Float? = null,
    val worldVelocityY: Float? = null,
    val worldVelocityZ: Float? = null,
    val gForceLateral: Float? = null,
    val gForceLongitudinal: Float? = null,
    val gForceVertical: Float? = null,
    val yaw: Float? = null,
    val pitch: Float? = null,
    val roll: Float? = null,
)

data class PacketMotionData(
    val carMotionData: List<CarMotionData> = emptyList(),
)

data class LapData(
    val position: Int? = null,
    val currentLapNum: Int? = null,
    val lapDistance: Float? = null,
    val lastLapTime: Float? = null,
    val bestLapTime: Float? = null,
    val sector1TimeMs: Int? = null,
    val sector2TimeMs: Int? = null,
    val currentSector: Int? = null,
    val deltaToCarInFront: Int? = null,
    val deltaToRaceLeader: Int? = null,
    val pitStatus: Int? = null,
    val driverStatus: Int? = null,
    val resultStatus: Int? = null,
    val numPitStops: Int? = null,
    val pitStopTimer: Float? = null,
    val pitLaneTime: Int? = null,
    val penalties: Int? = null,
    val totalWarnings: Int? = null,
    val cornerCuttingWarnings: Int? = null,
    val numUnservedDriveThroughs: Int? = null,
    val numUnservedStopGoPenalties: Int? = null,
    val lapInvalid: Boolean? = null,
)

data class PacketLapData(
    val lapData: List<LapData> = emptyList(),
)

data class CarSetupData(
    val frontWing: Int? = null,
    val rearWing: Int? = null,
    val onThrottle: Int? = null,
    val diffOnThrottle: Int? = null,
    val offThrottle: Int? = null,
    val frontCamber: Float? = null,
    val rearCamber: Float? = null,
    val frontToe: Float? = null,
    val rearToe: Float? = null,
    val frontSuspension: Int? = null,
    val rearSuspension: Int? = null,
    val frontAntiRollBar: Int? = null,
    val rearAntiRollBar: Int? = null,
    val frontSuspensionHeight: Int? = null,
    val rearSuspensionHeight: Int? = null,
    val brakePressure: Int? = null,
    val brakeBias: Int? = null,
    val rearLeftTyrePressure: Float? = null,
    val rearRightTyrePressure: Float? = null,
    val frontLeftTyrePressure: Float? = null,
    val frontRightTyrePressure: Float? = null,
    val ballast: Int? = null,
    val fuelLoad: Float? = null,
)

data class PacketCarSetupData(
    val carSetups: List<CarSetupData> = emptyList(),
)

data class CarTelemetryData(
    val speed: Int? = null,
    val throttle: Float? = null,
    val brake: Float? = null,
    val steering: Float? = null,
    val gear: Int? = null,
    val brakeTemp: List<Int> = listOf(0, 0, 0, 0),
    val tireSurfaceTemp: List<Int> = listOf(0, 0, 0, 0),
    val tireInnerTemp: List<Int> = listOf(0, 0, 0, 0),
    val tirePressure: List<Float> = listOf(0f, 0f, 0f, 0f),
    val surfaceType: List<Int> = listOf(0, 0, 0, 0),
    val engineRpm: Int? = null,
    val drs: Int? = null,
)

data class PacketCarTelemetryData(
    val carTelemetryData: List<CarTelemetryData> = emptyList(),
)

data class CarStatusData(
    val tractionControl: Int? = null,
    val antiLockBrakes: Int? = null,
    val fuelMix: Int? = null,
    val brakeBias: Int? = null,
    val pitLimiterStatus: Int? = null,
    val fuelInTank: Float? = null,
    val fuelCapacity: Float? = null,
    val fuelRemainingLaps: Float? = null,
    val maxRpm: Int? = null,
    val idleRpm: Int? = null,
    val maxGears: Int? = null,
    val drsAllowed: Int? = null,
    val actualTyreCompound: Int? = null,
    val visualTyreCompound: Int? = null,
    val tireAge: Int? = null,
    val vehicleFiaFlags: Int? = null,
    val ersStoreEnergy: Float? = null,
    val ersDeployMode: Int? = null,
    val ersHarvestedMGUK: Float? = null,
    val ersHarvestedMGUH: Float? = null,
    val ersDeployedThisLap: Float? = null,
)

data class PacketCarStatusData(
    val carStatusData: List<CarStatusData> = emptyList(),
)

data class CarDamageData(
    val tireWear: List<Int> = listOf(0, 0, 0, 0),
    val tiresDamage: List<Int> = listOf(0, 0, 0, 0),
    val brakesDamage: List<Int> = listOf(0, 0, 0, 0),
    val frontLeftWingDamage: Int? = null,
    val frontRightWingDamage: Int? = null,
    val rearWingDamage: Int? = null,
    val floorDamage: Int? = null,
    val diffuserDamage: Int? = null,
    val sidepodDamage: Int? = null,
    val gearboxDamage: Int? = null,
    val engineDamage: Int? = null,
)

data class PacketCarDamageData(
    val carDamageData: List<CarDamageData> = emptyList(),
)

data class LapHistoryData(
    val lapTimeInMs: Long? = null,
    val sector1TimeInMs: Int? = null,
    val sector2TimeInMs: Int? = null,
    val sector3TimeInMs: Int? = null,
    val lapValidBitFlags: Int? = null,
)

data class PacketSessionHistoryData(
    val carIndex: Int? = null,
    val numLaps: Int? = null,
    val bestLapTimeLapNum: Int? = null,
    val bestSector1LapNum: Int? = null,
    val bestSector2LapNum: Int? = null,
    val bestSector3LapNum: Int? = null,
    val bestSector1Ms: Int? = null,
    val bestSector2Ms: Int? = null,
    val bestSector3Ms: Int? = null,
    val lapHistoryData: List<LapHistoryData> = emptyList(),
)

data class TyreSetData(
    val actualTyreCompound: Int? = null,
    val visualTyreCompound: Int? = null,
    val wear: Int? = null,
    val available: Boolean? = null,
    val recommendedSession: Int? = null,
)

data class PacketTyreSetsData(
    val carIndex: Int? = null,
    val fittedIdx: Int? = null,
    val availableSets: List<TyreSetData> = emptyList(),
)
